\documentclass{article}
\usepackage[utf8]{inputenc}
\usepackage{amsmath}
\usepackage{geometry}
\usepackage{graphicx}
\usepackage[T1]{fontenc}
\usepackage{fourier}

\title{BIM3007-Assignment4}
\author{Junyang Deng}
\date{\today}
\geometry{left=3cm,right=3cm,top=2cm,bottom=2cm}
\setlength{\parindent}{0cm}
\linespread{1.3}
\begin{document} 
\maketitle
\section{RNA-Seq}
\begin{enumerate}
    \item In RNA-seq analysis, what is the significance of data quality control? What softwares can be used for quality control? \emph{(10 points)}
    
    \textbf{Significance}: Quality control is usually the first step in any RNA-seq analysis because it helps ensure the reliability of downstream analysis. Specifically, quality control can improve the accuracy of alignment and make differential analysis or functional annotation more reliable.

    \textbf{Softwares} for quality control include FastQC, RSeQC, MultiQC, RNASeqQC etc. In tutorial we learned about \textbf{FastQC} and \textbf{MultiQC}.
    
    \begin{itemize}
        \item FastQC: When a \texttt{.fastq} file is input, this software will generate a report for the data in \texttt{.html} format. The report includes the following tests and results (Figure~\ref{fig:fastqc}). The mark in front of each item indicates the test result. A green tick means `pass', an exclamation mark means `warning', and a red cross means `failure'.
    \begin{figure}[htbp]
        \centering
        \includegraphics[width=0.3\linewidth]{images/fastqc.png}
        \caption{Fastqc report}
        \label{fig:fastqc}
    \end{figure}
    \item MultiQC: This software can be used to combine the results from multiple QC tools into a single report. It allows you to easily compare the results from different samples or different analysis pipelines.
\end{itemize}

    \item After performing the data quality control for the RNA-Seq reads, we need to align the reads to the reference genome. What is the alignment program and what is the purpose of the alignment? \emph{(10 points)}
    
    \textbf{Significance}: Alignment helps people identify the corresponding gene of the transcript. With alignment information, people can use RNA-seq data to quantify the expression level of each gene or identify alternative splicing. 
    
    \textbf{Softwares} for alignment can be divided into two types: Splice-aware aligners (STAR, TopHat, HiSat) and non-splice aligners (BWA, Bowtie and HiSat). 
    \textbf{Splice-aware aligners} are willing to open gaps rather than mismatches when two sequences cannot coincide. On the contrary, \textbf{non-splice-aware aligners} are more likely to generate mismatches, and gaps usually have higher penalties. The splice-aware ones can be used when transcripts are mapped to the whole genome, while the non splice-aware ones can be used to map 

    \item After alignment step, we obtain bam files. What program help us to summarize the gene counts? \emph{(10 points)}
    
    Program: featureCounts. This program will generate a table of genes by samples with raw sequence abundances.
    \begin{figure}[htbp]
        \centering
        \includegraphics[width=0.8\linewidth]{images/feature-count.png}
        \caption{An example of table generated by featureCounts.}
        \label{fig:feature}
    \end{figure}

    \item What is the input and output of the standard RNA-seq analysis (featureCounts program analysis as the end point), and what kind of analysis can we do on its output? \emph{(10 points)}
    
    Input: \texttt{.fastq} file\\
    Output: a read count matrix\\
    Analysis: Many R packages can be used to conduct downstream analysis. For differential expressed gene (DEG) analysis can be done using DESeq2, limma and edgeR. After finding DEGs, we can do functional analysis using EnrichR and GSEA.

\end{enumerate}

\section{PTM Prediction Preprocessing}
\begin{enumerate}
    \item Please write a python script to extract the positive dataset and the negative dataset (13-mer) from the raw dataset (\texttt{Ubiquitination\_sites.txt}), and use CD-HIT program with cut-off values of sequence similarity of 80\% to remove the homologous sequences of both datasets. Please provide a table for the data statistics of positive and negative sequences before and after removing the homologous sequences. (30 points)
    
    After processing, 8000 positive and 39292 negative sequences with window size 13 are generated. 
    \item After the removal of homologous sequences, please apply WebLogo tool to generate sequence logos for positive sequences and negative sequences, respectively. Then, please employ TwoSampleLogo tool to conduct the comparison of position-specific AAC between positive and negative data. (15 points)
    \item Finally, Please give out a comparison of AAC (Amino acid composition) between positive and negative sequences in terms of bar-chart visualization. (15 points)
\end{enumerate}
\end{document}